### ISSUE TITLE
[QUERY] ServiceBusQueueOperation and ServiceBusTopicOperation extends deprecated interface (SendOperation)

### ISSUE THREAD

**Issue by taisuke-fujimoto** [Azure/azure-sdk-for-java#26923]
**Query/Question**
* com.azure.spring.integration.servicebus.queue.ServiceBusQueueOperation
* com.azure.spring.integration.servicebus.topic.ServiceBusTopicOperation

I'm using these interfaces for sending to the Service Bus.
And I found that the interface (SendOperation) they inherit are deprecated.

Is there any alternative to avoid using deprecations? 

**Setup (please complete the following information if applicable):**
 - Library/Libraries: com.azure.spring:azure-spring-integration-servicebus:2.13.0

**Comment by joshfree** [comment]
@yiliuTo could you please take a look?

**Comment by yiliuTo** [comment]
Hi @taisuke-fujimoto , the deprecated class of `SendOperation` cannot be replaced currently, and you could replace the `ServiceBusQueueOperation` and `ServiceBusTopicOperation` in our 4.x version with the class of `com.azure.spring.servicebus.core.ServiceBusTemplate` for sending to either queue or topic, for the 4.x version, you can migrate to library of [spring-cloud-azure-starter-integration-servicebus](https://mvnrepository.com/artifact/com.azure.spring/spring-cloud-azure-starter-integration-servicebus). 

For more details about migration to 4.x, you can refer to the [migration guide](https://microsoft.github.io/spring-cloud-azure/4.0.0-beta.3/4.0.0-beta.3/reference/html/appendix.html#migration-guide-for-4-0), [reference doc](https://microsoft.github.io/spring-cloud-azure/4.0.0-beta.3/4.0.0-beta.3/reference/html/index.html#spring-integration-with-azure-service-bus) and [samples](https://github.com/Azure-Samples/azure-spring-boot-samples/tree/spring-cloud-azure_4.0.0-beta.3/servicebus/spring-cloud-azure-starter-integration-servicebus).

Besides, if you plan to use the ServiceBusTemplate to send messages directly instead of the Spring Integration framework, then you could call the `setDefaultEntityType` API for the entity type and then call the related sendAsync APIs as 2.13.0.

**Comment by taisuke-fujimoto** [comment]
@yiliuTo 
Thank you for the information about the 4.x version.
I will migrate when the stable 4.x version is released.

> Besides, could you share how you use the ServiceBusQueueOperation/ServiceBusTopicOperation?

I'm using the Spring framework.
I use the ServiceBusQueueOperation/ServiceBusTopicOperation like [this sample code](https://github.com/Azure-Samples/azure-spring-boot-samples/tree/71c87bb467be3f0f3dee421fbcb3607350f6563d/servicebus/azure-spring-cloud-starter-servicebus/servicebus-operation).

**Comment by yiliuTo** [comment]
Hi @taisuke-fujimoto , thanks for your information. And due to we are going to migrate to Reactor, we will not support CompletableFuture in the 4.x version. And for 2.x libraries, we will not provide other undeprecated alternatives as mentioned above. For the stable 4.x version, you can subscribe the issue https://github.com/Azure/azure-sdk-for-java/issues/24327 and we will also notice you when 4.0.0 get released.

Besides, if you use the ServiceBusQueueOperation/ServiceBusTopicOperation for sending messages directly, then in the 4.x version, you can also choose to use our autoconfiguration of the `ServiceBusClientBuilder` bean to build multiple `ServiceBusSenderAsyncClient`\`ServiceBusSenderClient` with different destinations as you need and then send messages in the Service Bus Java Client SDK way, which is provided in the library of [com.azure.spring: spring-cloud-azure-starter-servicebus](https://mvnrepository.com/artifact/com.azure.spring/spring-cloud-azure-starter-servicebus)

**Comment by taisuke-fujimoto** [comment]
Thanks, I subscribed to the issue #24327

**Comment by yiliuTo** [comment]
Hi @taisuke-fujimoto , the library of com.azure.spring:spring-cloud-azure-starter-integration-servicebus:4.0.0 has been released, you can refer to https://search.maven.org/artifact/com.azure.spring/spring-cloud-azure-starter-integration-servicebus.
Besides, here are the [migration guide](https://microsoft.github.io/spring-cloud-azure/current/reference/html/appendix.html#_from_azure_spring_cloud_starter_servicebus_to_spring_cloud_azure_starter_integration_servicebus) and [reference doc](https://microsoft.github.io/spring-cloud-azure/current/reference/html/index.html#spring-integration-with-azure-service-bus).

**Comment by taisuke-fujimoto** [comment]
@yiliuTo thanks for the notice.

I have a question about `ServiceBusTemplate`.

I tested queue and topic sending using `ServiceBusTemplate`. (without using `@MessagingGateway`)
this is the code: https://github.com/taisuke-fujimoto/spring-azure-servicebus-sample
The test was successful even if I set the wrong `ServiceBusTemplate.defaultEntityType`.

Is it correct that `ServiceBusTemplate.defaultEntityType` does not affect the sending of queue and topic?
In other words, isn't it necessary to separate the `ServiceBusTemplate` instance by queue and topic?

As a background, I saw [this sample code](https://github.com/Azure-Samples/azure-spring-boot-samples/tree/spring-cloud-azure_4.0.0/servicebus/spring-cloud-azure-starter-integration-servicebus/single-namespace) and thought it strange.
Because it is using same `ServiceBusTemplate` instance for queue sending and topic sending.

**Comment by jialigit** [comment]
> I tested queue and topic sending using `ServiceBusTemplate`. (without using `@MessagingGateway`)

So, this is a new question, right? I will take a look at it. @yiliuTo , is this really a problem? The ServiceBusTemplate.
And what about the original problem, you have any more questions about it? Thanks for your reply.

**Comment by yiliuTo** [comment]
@jialigit you can have a try with behavior @taisuke-fujimoto  mentioned to see whether it exists

**Comment by jialigit** [comment]
> @jialigit you can have a try with behavior @taisuke-fujimoto mentioned to see whether it exists

Yes, I will take a took a look.

**Comment by jialigit** [comment]
@yiliuTo @taisuke-fujimoto 
Yes, the ServiceBusTemplate may have a confusing method here. 
```
 ServiceBusSenderAsyncClient sender = new ServiceBusClientBuilder()
     .connectionString(connectionString)
     .sender()
     .queueName(queueName)
     .buildAsyncClient();

 // Creating a batch without options set, will allow for automatic routing of events to any partition.
 sender.createMessageBatch().flatMap(batch -> {
     batch.tryAddMessage(new ServiceBusMessage(BinaryData.fromBytes("test-1".getBytes(UTF_8))));
     batch.tryAddMessage(new ServiceBusMessage(BinaryData.fromBytes("test-2".getBytes(UTF_8))));
     return sender.sendMessages(batch);
 }).subscribe(unused -> {
 },
     error -> System.err.println("Error occurred while sending batch:" + error),
     () -> System.out.println("Send complete."));
```

In the above snippet, queueName or topicName make no difference.

**Comment by jialigit** [comment]
For the original query is given the answer for it. And the new question mentioned in the post is tracked in another issue.  we close this issue now.

**Comment by taisuke-fujimoto** [comment]
@jialigit 
Thanks for the information.
I think I understand about the ServiceBusSenderAsyncClient and the Service Bus Namespace.

For now, one ServiceBusTemplate instance can send both queue and topic (whatever the value of `defaultEntityType` is).

I hope the `ServiceBusTemplate.defaultEntityType` and `ServiceBusSenderAsyncClient.queueName / topicName` will be refactored in the future.

**Comment by jialigit** [comment]
> @jialigit Thanks for the information. I think I understand about the ServiceBusSenderAsyncClient and the Service Bus Namespace.
> 
> For now, one ServiceBusTemplate instance can send both queue and topic (whatever the value of `defaultEntityType` is).
> 
> I hope the `ServiceBusTemplate.defaultEntityType` and `ServiceBusSenderAsyncClient.queueName / topicName` will be refactored in the future.

@taisuke-fujimoto Thank you. For now, we track this issue of refactoring by this: #28125, pending for the author team response.

**Comment by jialigit** [comment]
@taisuke-fujimoto , Hi! As the developing team respond, I also think that makes sense. Would you please directly track this issue directly in the thread: #28125 if you have new ideas?


<img width="623" alt="image" src="https://user-images.githubusercontent.com/13768484/163306888-208e28ab-221f-4f09-a7d6-65b073be311d.png">

**Comment by jialigit** [comment]
@taisuke-fujimoto Thank you for your attention to the new issue to track the problem(#28125). We may close this issue or, if you have new ideas, let's reopen it. Thank you again.

**Comment by taisuke-fujimoto** [comment]
@jialigit Hi, My new idea is to split the ServiceBusTemplate into ServiceBusQueueTemplate and ServiceBusTopicTemplate (as in previous versions).

Because:
* ServiceBusTopicTemplate.defaultEntityType may be a required value in the future (even if it doesn't make any sense now), so you shouldn't remove this.
  However, by splitting the ServiceBusTemplate into ServiceBusQueueTemplate and ServiceBusTopicTemplate, the defaultEntityType can be removed from the split class.
* Considering the Spring Bean that the library automatically registers, it is better to define the ServiceBusTemplate class of the queue and the topic differently.
  With the current library, it is a bit annoying because it is necessary to set the defaultEntityType of the automatically registered ServiceBusTemplate.

**Comment by jialigit** [comment]
@taisuke-fujimoto Thanks very much for your suggestion. I think ServiceBusTemplate is short and convenient while split  it into two "templates" is also a good option. But that will also bring new complexities and breaking changes. As a balance, we may not put this refactor plan on the list in the future.
